### PR TITLE
cross-compile container images

### PR DESCRIPTION
**Pull request by ereslibre**
Create multi-arch container images by first cross compiling the
binaries for the different platforms, and then creating the container
images by copying the binaries.

### CODE REVIEW

**Comment by ereslibre** [comment]
Converting to draft because of the usage of `musl.cc`.

File: .github/workflows/container-image.yml [hunk 1]
```diff
@@ -4,54 +4,106 @@ on:
...
+      - name: Build policy-server
+        env:
+          CC: x86_64-linux-musl-gcc
+        run: |
+          cargo build --target=x86_64-unknown-linux-musl --release
+      - name: Upload policy-server
+        uses: actions/upload-artifact@v2
+        with:
+          name: policy-server-x86_64
+          path: target/x86_64-unknown-linux-musl/release/policy-server
```
Comment (flavio) [inline .github/workflows/container-image.yml:34]: I guess this why you changed the `.Dockerignore` to allow the `target` directory. I would just copy this file (and the aarch64 one) into the root of the git checkout. Obviously using two different names (like `policy-server-x86_64`)

**Review by flavio — COMMENTED** [review]
Cool stuff :clap:

**Review by viccuad — COMMENTED** [review]
Thanks for this, this is great! I'm amazed that the compilation time went down to 15 minutes.

Did we talk about performance of running code already? I have read that Musl is not as efficient as libc on mem allocation. I suppose not a problem for our usage, but maybe worth checking.

With this, we will be moving our targets from Rust Tier 1 (guaranteed to work) to [Tier 2](https://doc.rust-lang.org/rustc/platform-support.html#tier-2-with-host-tools) (guaranteed to build). For x86_64 this means a reduction in assurances. I suppose it's not a problem, but something to also have in mind.

File: .cargo/config [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[target.x86_64-unknown-linux-musl]
+linker = "x86_64-linux-musl-ld"
+ar = "x86_64-linux-musl-ar"
+
+[target.aarch64-unknown-linux-musl]
+linker = "aarch64-linux-musl-ld"
+ar = "aarch64-linux-musl-ar"
```
[the remark this answers is not in the document]
Comment (ereslibre) [reply .cargo/config:7]: Without it, `ring`'s `build.rs` (https://github.com/briansmith/ring/blob/main/build.rs) fails to cross-compile.

Related: https://github.com/rust-lang/cargo/issues/4133

File: Dockerfile [hunk 3]
```diff
@@ -1,32 +1,21 @@
...
+COPY --from=common /etc/group /etc/group
+COPY --chmod=0755 policy-server-aarch64 /
 
+# common final steps
+FROM build-${TARGETARCH}
+USER 65533:65533
 EXPOSE 3000
-
-ENTRYPOINT ["/usr/local/bin/policy-server"]
+ENTRYPOINT ["/policy-server"]
```
Comment (viccuad) [inline Dockerfile:21]: Silly question, is there a reason to put the bin out of path?

**Review by viccuad — APPROVED** [review]
LGTM! yay!
Do we want to cross-compile on development too?
I tried here `cargo build --target x86_64-unknown-linux-musl` and it failed on the ring issue, so from what I understand it's not using `.cargo/config`.

File: .cargo/config (hunk 2 above)
[the remark this answers is not in the document]
Comment (flavio) [reply .cargo/config:12]: Oh, that explains the issues I've seen with musl compilation with ring (even without cross build). Thanks

**Review by flavio — COMMENTED** [review]
Overall great :clap: 

I left some comments here and there

**Comment by flavio** [comment]
> Do we want to cross-compile on development too? I tried here `cargo build --target x86_64-unknown-linux-musl` and it failed on the ring issue, so from what I understand it's not using `.cargo/config`.

Actually that would be nice, just to be able to reproduce bugs locally...

File: .github/workflows/container-image.yml [hunk 4]
```diff
@@ -4,54 +4,104 @@ on:
...
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Setup rust toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+      - run: rustup target add x86_64-unknown-linux-musl
+      - name: Setup musl for x86_64
+        run: |
+          curl -O https://musl.cc/x86_64-linux-musl-cross.tgz
```
[the remark this answers is not in the document]
Comment (ereslibre) [reply .github/workflows/container-image.yml:22]: I also think this is a problem. I see two possible options:

1. Upload the tarball used for a release somewhere (along with the release?). This does not seem "pretty". Another issue is data contention on hosted platforms: for how long would that tarball be kept -- e.g. had we uploaded it to GitHub Actions? If we upload it as a release asset I think data contention should be take for granted, but it's weird to show final binaries and the libc along with them -- maybe we could somehow add the libc to the source code tarball that is part of the release?
2. Use a reproducible package manager on top, e.g. nix. Everyone on the team (and the CI) would always refer to the same versions of all dependencies.

I think it's not a blocker **right now**, but I completely see this as a blocker for a first stable and supported reelase.

File: Dockerfile (hunk 3 above)
Comment (ereslibre) [reply Dockerfile:21]: We can put it under any `bin` dir but I honestly don't see a strong reason to do so. The tree would be as @flavio described.

File: Dockerfile [hunk 5]
```diff
@@ -1,32 +1,21 @@
...
 RUN echo "policy-server:x:65533:65533::/tmp:/sbin/nologin" >> /etc/passwd
-# Add the default GID to /etc/group for completeness.
 RUN echo "policy-server:x:65533:policy-server" >> /etc/group
 
-COPY --from=builder /usr/local/cargo/bin/policy-server /usr/local/bin/policy-server
+# amd64-specific
+FROM scratch AS build-amd64
+COPY --from=common /etc/passwd /etc/passwd
+COPY --from=common /etc/group /etc/group
+COPY --chmod=0755 policy-server-x86_64 /
```
[the remark this answers is not in the document]
Comment (ereslibre) [reply Dockerfile]: Good catch. Fixed.

**Comment by ereslibre** [comment]
>> Do we want to cross-compile on development too? I tried here cargo build --target x86_64-unknown-linux-musl and it failed on the ring issue, so from what I understand it's not using .cargo/config.

> Actually that would be nice, just to be able to reproduce bugs locally...

I also agree. I think is a good strategy. I would push for that right after this work. We can discuss about it. My preferred mechanism would be to have this as a nix project, so our dependencies are pinned and easily built (no need to fill pre-requirements prior to build other than having nix installed).

I think is worth checking, let me know if that's good with you and if you think we should plan for that soon.

**Comment by ereslibre** [comment]
> Do we want to cross-compile on development too?

You can manually do this already, but is not pretty. Let me know if this fixes the build & link issue for you.

```
$ curl https://musl.cc/x86_64-linux-musl-cross.tgz | tar -xz
$ export PATH=$PWD/x86_64-linux-musl-cross/bin:$PATH
$ cargo build --target x86_64-unknown-linux-musl
```

Same for targeting aarch64 from x86_64:

```
$ curl https://musl.cc/aarch64-linux-musl-cross.tgz | tar -xz
$ export PATH=$PWD/aarch64-linux-musl-cross/bin:$PATH
$ cargo build --target aarch64-unknown-linux-musl
```

This should work AFAICT.

File: Dockerfile (hunk 3 above)
Comment (viccuad) [reply Dockerfile:21]: doh true, using scratch indeed. Don't mind, then :).

File: .github/workflows/container-image.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (flavio) [reply .github/workflows/container-image.yml]: Yes, this is not a blocker right now. let's look into that later on

File: .github/workflows/container-image.yml [hunk 6]
```diff
@@ -1,57 +1,102 @@
...
+        run: |
+          curl https://musl.cc/x86_64-linux-musl-cross.tgz | tar -xz
+          echo "$PWD/x86_64-linux-musl-cross/bin" >> $GITHUB_PATH
+      - run: touch policy-server-x86_64
+      # - name: Build policy-server
+      #   env:
+      #     CC: x86_64-linux-musl-gcc
+      #   run: |
+      #     cargo build --target=x86_64-unknown-linux-musl --release
+      #     mv target/x86_64-unknown-linux-musl/release/policy-server policy-server-x86_64
```
Comment (flavio) [inline .github/workflows/container-image.yml:25]: This smells like some debugging going on :)

File: .github/workflows/container-image.yml [hunk 7]
```diff
@@ -1,57 +1,102 @@
...
+        run: |
+          curl https://musl.cc/aarch64-linux-musl-cross.tgz | tar -xz
+          echo "$PWD/aarch64-linux-musl-cross/bin" >> $GITHUB_PATH
+      - run: touch policy-server-aarch64
+      # - name: Build policy-server
+      #   env:
+      #     CC: aarch64-linux-musl-gcc
+      #   run: |
+      #     cargo build --target=aarch64-unknown-linux-musl --release
+      #     mv target/aarch64-unknown-linux-musl/release/policy-server policy-server-aarch64
```
Comment (flavio) [inline .github/workflows/container-image.yml:52]: Same here

**Review by flavio — COMMENTED** [review]
I think there's still some work going on, am I right? This doesn't look ready to be merged

File: .github/workflows/container-image.yml (hunk 6 above)
Comment (ereslibre) [reply .github/workflows/container-image.yml]: Yes, I am still checking that the image can be built properly on my fork. I will ping you here when the build has succeeded in Github actions in my fork.

**Comment by ereslibre** [comment]
> I think there's still some work going on, am I right? This doesn't look ready to be merged

This is mergeable. I was testing GitHub behaviors when building the container image. Certain parts are a bit of trial and error. It's good now.

**Review by jvanz — APPROVED** [review]
LGTM! :D

**Review by flavio — APPROVED** [review]
Great, go ahead with it :clap:

**Pull request merged**